### ISSUE TITLE
Progress on thread safety?

### ISSUE THREAD

**Issue by systemed** [LLNL/zfp#17]
I'm experimenting with using zfp to store terrain (elevation) data. Early results are very encouraging, saving 50% over my current compression solution - thank you.

The project I'm building it into is multi-threaded (for read access only). Before I go any further with integrating it, could I ask if there's any progress on making zfp thread-safe?

**Comment by lindstro** [comment]
We have an experimental feature branch (feature/view) that supports thread-safe compressed arrays via "private views."  For read-only access, each thread may access any part of the array, which is backed by a thread-private cache (this avoids expensive locking).  For read-write access, the array must be partitioned into independent zfp blocks (e.g. blocks of 4x4 values in 2D) via non-overlapping read-write views, and each thread may access only the blocks assigned to it (whether reading or writing).  No cache coherence is enforced, so it's up to each thread to flush its cache to ensure consistency.  See examples/diffusion.cpp for an example of both read-only and read-write private views.

Note that we still need to work out some performance issues, and it is possible that we'll make some minor API changes before this feature is officially released in September.  Any feedback would be appreciated.

**Comment by systemed** [comment]
This is terrific - thank you. I've experimented with the view branch and it seems to work well. I don't anticipate I'll be integrating it until September anyway so look forward to the next release.

**Comment by systemed** [comment]
Just to let you know I'm now using ZFP in production for the elevation graphs at https://cycle.travel/map (plot a route, click the 'elevation' button). Thanks for all your work!

**Comment by lindstro** [comment]
That's great to hear.  So zfp is being used to store the 2D elevation map, z(x, y), in compressed form and not just the 1D elevation vs. distance graph being shown for a given route?

**Comment by systemed** [comment]
Exactly that. It's compiled with -DZFP_BIT_STREAM_WORD_SIZE=16, and integrated into a patched version of OSRM (http://project-osrm.org).

**Comment by lindstro** [comment]
Neat.  Is this patch public and something that others can benefit from?  I'm not familiar with OSRM...

I can't help but ask about the choice of a 16-bit word size, which specifies how many bits of compressed data to read or write at a time.  Most users would use a word size of 8 (to ensure compatibility between little and big endian machines) or 64 (the default, for maximum speed).  Why did you decide on 16 bits?  I'm asking to get a better understanding of user needs.  Supporting multiple word sizes causes compatibility problems, and we have been debating whether or not to hide this parameter from the user.

**Comment by systemed** [comment]
Minimising RAM usage is crucial for me - I keep elevation for the whole of Europe and North America memory-resident, plus a routing graph with [Contraction Hierarchies](https://en.wikipedia.org/wiki/Contraction_hierarchies) which are pretty memory-hungry. So I was anxious to find the sweet spot between good-enough results and saving memory.

After a bit of experimentation, I've found that fixed rate 4 is good for computing the routing graph, and I can drop down to fixed rate 2 for the on-demand elevation profiles (where memory is most constrained). I did try to use 2 for computing the routing graph, but on rare occasions it would be out by about 10m elevation, which resulted in an unduly harsh climb penalty and the router refusing to take that road.

So the 16-bit word size (AIUI) allows me to choose those rates while maintaining a small performance advantage over uint8. The routing graph is computed on, and the routing served from, x86 machines running Ubuntu so there's no compatibility issue.

I've not contributed the patch upstream to OSRM because I'm running a fork from an older version (OSRM trunk has since diverged towards car navigation use-cases), but I've been talking to a couple of people in the OpenStreetMap community who use elevation data, and so what I'll probably do is make my tooling public so that others can adapt it to their needs. That said, the future of OSRM is a little uncertain at the moment so I wouldn't necessarily rule it out!
